From 637e09560752460a6e97ba79e9a585bdc87d5f8d Mon Sep 17 00:00:00 2001
From: Christian Lindig <christian.lindig@cloud.com>
Date: Fri, 11 Apr 2025 15:31:48 +0100
Subject: [PATCH] rrd_file_writer: protect against resource leak

Make sure we always close the file descriptor on the error paths.

Signed-off-by: Christian Lindig <christian.lindig@cloud.com>
---
 .../lib/transport/file/rrd_file_writer.ml     | 19 ++++++++++++-------
 1 file changed, 12 insertions(+), 7 deletions(-)

diff --git a/ocaml/xcp-rrdd/lib/transport/file/rrd_file_writer.ml b/ocaml/xcp-rrdd/lib/transport/file/rrd_file_writer.ml
index 9f8fa196d..92cff3e49 100644
--- a/ocaml/xcp-rrdd/lib/transport/file/rrd_file_writer.ml
+++ b/ocaml/xcp-rrdd/lib/transport/file/rrd_file_writer.ml
@@ -14,6 +14,8 @@
 
 type local_id = {path: string; shared_page_count: int}
 
+let finally f finally = Fun.protect ~finally f
+
 module File = struct
   let page_size = 4096
 
@@ -29,14 +31,17 @@ module File = struct
   let init {path; shared_page_count} =
     let size = shared_page_count * page_size in
     let fd = Unix.openfile path [Unix.O_RDWR; Unix.O_CREAT] 0o600 in
-    let mapping =
-      Bigarray.(
-        array1_of_genarray @@ Unix.map_file fd char c_layout true [|size|]
+    finally
+      (fun () ->
+        let mapping =
+          Bigarray.(
+            array1_of_genarray @@ Unix.map_file fd char c_layout true [|size|]
+          )
+        in
+        let cstruct = Cstruct.of_bigarray mapping in
+        (path, cstruct)
       )
-    in
-    Unix.close fd ;
-    let cstruct = Cstruct.of_bigarray mapping in
-    (path, cstruct)
+      (fun () -> Unix.close fd)
 
   let cleanup _ path _ = Unix.unlink path
 
